### ISSUE TITLE
Text and image association bug

### ISSUE THREAD

**Issue by Nenotriple** [DevArqSangoi/das-EzBooruTagEditor#2]
The wrong filename is sometimes being incorrectly associated with the image/tags displayed. However, the correct img-txt pair is being loaded and the correct txt file is being saved.

![1](https://github.com/DevArqSangoi/das-EzBooruTagEditor/assets/70049990/d361c29d-0be9-4dc2-a3c0-7edcfaa39745)

![2](https://github.com/DevArqSangoi/das-EzBooruTagEditor/assets/70049990/a4779c70-6297-4679-8e2d-be666fc01b98)

![3](https://github.com/DevArqSangoi/das-EzBooruTagEditor/assets/70049990/60bb60e3-b2f1-40e3-9f71-a49a82824bc5)

Here are the image and text files used in this test.
[test.zip](https://github.com/DevArqSangoi/das-EzBooruTagEditor/files/13434923/test.zip)

The issue appears to be in the `show_file_content` function. Because the order of files in the `listbox` and the order of files in the `file_map` are different.

Instead of using the `selected_index` from the `listbox` to get the file from the `file_map`, you should get the filename directly from the `listbox`.

So change this:

```

    def show_file_content(self, event):
        self.currently_marked_tag = []
        selected_index = self.listbox.curselection()
        if not selected_index:
            return

        self.current_file = list(self.file_map.keys())[selected_index[0]]
        tag_file_path = os.path.join(self.current_directory, self.current_file)

        self.unique_tags, self.common_words_tags = self.read_tags_from_file(tag_file_path)
        self.removed_tags = []

        self.tags = self.unique_tags.copy()
        for tags in self.common_words_tags.values():
            self.tags.extend(tags)

        self.create_tag_widget()
        self.resize_image()

```

To this:

```

    def show_file_content(self, event):
        self.currently_marked_tag = []
        selected_index = self.listbox.curselection()
        if not selected_index:
            return

        self.current_file = self.listbox.get(selected_index) ### This line. ###
        tag_file_path = os.path.join(self.current_directory, self.current_file)

        self.unique_tags, self.common_words_tags = self.read_tags_from_file(tag_file_path)
        self.removed_tags = []

        self.tags = self.unique_tags.copy()
        for tags in self.common_words_tags.values():
            self.tags.extend(tags)

        self.create_tag_widget()
        self.resize_image()

```

Some screenshots with the fix applied.

![1_fix](https://github.com/DevArqSangoi/das-EzBooruTagEditor/assets/70049990/dc77ef79-98f2-4d15-8611-27b2591ca977)

![2_fix](https://github.com/DevArqSangoi/das-EzBooruTagEditor/assets/70049990/32773640-eb1d-4ced-834b-aee31f136cc8)

![3_fix](https://github.com/DevArqSangoi/das-EzBooruTagEditor/assets/70049990/64b6aee6-c64f-4117-80be-4dc21256918f)

Not thoroughly tested, but it appears to work as expected now.

This was a tough/fun issue to sus out! I originally thought it was related to the `natural_sort_key` function and was running in circles fixing nothing. Eventually I added print statements to everything and found the `file_map` was correct, but there was a discrepancy with the `selected_index` and `self.current_file`.

**Comment by sangoi-exe** [comment]
Hello! Thanks for the report, I'll check it out here and adjust according to your suggestion!

I believe this must be the cause of a problem that another user on Reddit is having, where they say he delete the tags, save, but when he go back to the image, the tags are there again. Must be deleting tags from a different file without knowing.

I'll add some alerts and information to the interface as well, like file names and stuff.

**Comment by sangoi-exe** [comment]
@Nenotriple, could you please check the changes? I just pushed them.

I'm going to update the README with the changes, I added some new features;

I made some deeper changes in the listing issues, added an extra step of organizing the file_map in the show_file_content, and adjusted the resize_image function that I had forgotten after implementing a new sorting of the file list.

**Comment by Nenotriple** [comment]
@DevArqSangoi 

The changes you made to fix this issue work great.

I did find 1 bug related to image resizing. It happens while viewing a wide aspect ratio image, and resizing the window and "squishing" the image frame so the image disappears.

The error didn’t occur with tall aspect ratio images because the code adjusts the width and height values to be greater than 0 when the image is taller than the frame. `if new_height > frame_height:`

```

 Selected file: 40-0000.txt, Image path: D:/image/test\40-0000.webp
 Exception in Tkinter callback
 Traceback (most recent call last):
   File "C:\Users\egegi\AppData\Local\Programs\Python\Python310\lib\tkinter\__init__.py", line 1921, in __call__
     return self.func(*args)
   File "C:\Users\egegi\Downloads\EzBooruTagEditor.py", line 131, in <lambda>
     self.right_frame.bind("<Configure>", lambda event: self.resize_image())
   File "C:\Users\egegi\Downloads\EzBooruTagEditor.py", line 245, in resize_image
     image = image.resize((new_width, new_height))
   File "C:\Users\egegi\AppData\Local\Programs\Python\Python310\lib\site-packages\PIL\Image.py", line 2199, in resize
     return self._new(self.im.resize(size, resample, box))
 ValueError: height and width must be > 0

```

I happens because `resize_image` tries to resize the image to a width and height of 0, which is not allowed by PIL. the fix prevents this by checking that the new width and height are both greater than 0 before resizing.

```

    def resize_image(self):
        if not self.current_image_path:
            return

        image = Image.open(self.current_image_path)
        aspect_ratio = image.width / image.height

        self.middle_frame.update_idletasks()
        frame_width = self.middle_frame.winfo_width()
        frame_height = self.middle_frame.winfo_height()

        new_width = frame_width
        new_height = int(new_width / aspect_ratio)

        if new_height > frame_height:
            new_height = frame_height
            new_width = int(new_height * aspect_ratio)

        # Add a check to ensure new_width and new_height are > 0
        if new_width > 0 and new_height > 0:
            image = image.resize((new_width, new_height))
            self.current_image = ImageTk.PhotoImage(image)

        self.image_box.config(image=self.current_image)

```

I'd also suggest adding a `minsize` to keep the window from being squished beyond useable range. This size is fairly small and works good for square aspect ratio images.

```

if __name__ == "__main__":
    root = tk.Tk()
    app = TextImageEditor(root)
    root.geometry("800x600")
    root.minsize(800,350) # Add this line.

```

**Comment by Nenotriple** [comment]
I somehow missed this comment.

> I believe this must be the cause of a problem that another user on Reddit is having, where they say he delete the tags, save, but when he go back to the image, the tags are there again. Must be deleting tags from a different file without knowing.

In my testing, the correct image and tags were being loaded. So when the tags were written to the text file, they were actually being written to the correct file. I'm not sure if the issue is related or not, but it's certainly possible.

**Comment by sangoi-exe** [comment]
@Nenotriple done! I included the `root.minsize` in the code.

Btw, do you mind trying to reproduce this bug? I couldn't and can't even imagine how this is happening:
[Reddit post from r/StableDiffusion](https://www.reddit.com/r/StableDiffusion/comments/17zubkj/comment/kae51vr/?utm_source=reddit&utm_medium=web2x&context=3)

My guess is that he is using either the 1.0 or 1.1 version.

**Comment by Nenotriple** [comment]
I have not been able to reproduce that bug, I only tried on the latest version.

**Comment by sangoi-exe** [comment]
@Nenotriple ok! Thank you!

I just fixed a bug similar to the other one you reported, where the tag delete function would stop working when switching images with a token marked. I found out that it also stopped working if a tag was added while a token was marked. I simply relocated the line I had used to fix the first bug to a better place. 

The bugs essentially happened because the function to create the widgets was called without removing the marked token for deletion from the variable, so I placed a line at the beginning of the widget creation function to clear the variable.

I'm going to close this issue then. If anything comes up, just give me a shout! Thank you so much for your help!
